tensorflow/core/graph/mkl_optimizer_merge.h

```diff
@@ -0,0 +1,217 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// An optimization pass that performs node merging and rewrite on graph nodes
+
+#ifndef TENSORFLOW_CORE_GRAPH_MKL_OPTIMIZER_MERGE_H_
+#define TENSORFLOW_CORE_GRAPH_MKL_OPTIMIZER_MERGE_H_
+
+#ifdef INTEL_MKL
+
+#include <sys/types.h>
+#include <vector>
+#include <string>
+#include <memory>
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+
+namespace tensorflow {
+
+// How many hops do we search for matching node in the backward dataflow graph?
+// We use maxhop of 10 based on empirical observations. Also, these are
+// maxhops in backward data-flow graph. Since input of forward nodes (Conv2D)
+// directly goes to backward nodes, we do not expect the hop-distance
+// would be more than few nodes.
+#define NODEMERGE_CONTEXT_MAXDEPTH 10
+
+// This optimization pass performs two tasks: merge
+// nodes in the forward pass, and rewrite the gradient ops
+// corresponding to merged forward ops.
+//
+// Merging nodes in the graph: Currently, it merges Conv2D+AddBias
+// and MatMul+AddBias nodes together.
+//
+// Rewriting nodes in the graph:
+// This is neded in order to optimize gradient ops of
+// Conv2D+AddBias and MatMul+AddBias. Gradient op of
+// both the Conv2D and MatMul is BiasAddGrad, and we
+// need to rewrite BiasAddGrad into Conv2D-specific BiasAddGrad,
+// and MatMul-specific BiasAddGrad. This is context-specific
+// optimization, where the context is the forward operator
+// that the BiasAddGrad corresponds to.
+//
+class NodeMergeRewritePass : public GraphOptimizationPass {
```

Comment:
At a high-level, its not clear to me that all of the code in this header file is needed elsewhere?  If that is not the case, could you move all of this code to the .cc file, and just make a thin wrapper class here in the .h file that inherits from GraphOptimizationPass?  In the .cc file, the thin wrapper classes Run method could create an instance of NodeMergeRewriter and invoke it (so all NodeMergeRewrite class definition could be in .cc file).